Fix unpinned hdmf in caiman conda env breaking WF-05 e2e

### Problem

The Weekly Regression e2e lane fails on `develop-main` at `WF-05 - Run Tutorial2 workflow via RUN ALL` (167 passed, 1 failed, 3 attempts). Run: https://github.com/arayabrain/araya-optinist/actions/runs/32676365616

Playwright only reports `Tutorial2 recorded "error", not success`. The real cause is in the `POST /run/result/2/f3afd5d5` response bodies, recovered from the run's Playwright trace. The first node of the pipeline, `caiman_mc`, fails while writing its NWB file:

```
File "/app/studio/app/common/core/rules/runner.py", line 164, in __save_func_nwb
  save_nwb(
File "/app/studio/app/optinist/core/nwb/nwb_creater.py", line 35, in acquisition
  new_nwbfile = NWBFile(
File "/app/.snakemake/conda/81b69ce01f2592c0ca301949e978514a_/lib/python3.11/site-packages/hdmf/container.py", line 342, in __new__
  inst = super().__new__(cls)
TypeError: Can't instantiate abstract class NWBFile with abstract method external_resources
```

The other three node errors in the same response are cascade only, not independent failures:

```
caiman_cnmf_t28ssmgpcz: AssertionError: Invalid node input data content. [.../caiman_mc_o5k0ez4d9p/caiman_mc.pkl]
pca_cr3lvekylp:         AssertionError: Invalid node input data content. [.../caiman_cnmf_t28ssmgpcz/caiman_cnmf.pkl]
post_process_0:         AssertionError: Invalid node input data content. [.../pca_cr3lvekylp/pca.pkl]
```

Root cause: hdmf 5.0.0 made `HERDManager` an abstract interface with an abstract `external_resources` property, and subclasses must now declare it in `__fields__`. pynwb 2.x's `NWBFile` does not, so any pynwb 2 plus hdmf >= 5 pairing cannot instantiate an NWBFile.

Why that pairing got installed: `studio/app/optinist/wrappers/caiman/conda/caiman.yaml` pins `pynwb<3` but does not pin hdmf, and conda-forge `pynwb 2.8.3` declares only `hdmf >=3.14.5` with no upper bound. The snakemake wrapper envs are not pre-created in `studio/config/docker/Dockerfile`, so this env is solved fresh from conda-forge on every CI run and is free to move a major version.

Confirmed by a dry-run solve of the exact `caiman.yaml` spec set for linux-64 (`CONDA_OVERRIDE_GLIBC=2.17 mamba create --dry-run --platform linux-64`):

```
+ hdmf   6.2.0  pyhd8ed1ab_0  conda-forge
+ pynwb  2.8.3  pyhd8ed1ab_1  conda-forge
```

Timeline: WF-05 passed on the same repo content in runs 31981674833 (Aug 17, 5.7m) and 32089054683 (Aug 18, 5.0m). hdmf 6.2.0 landed on conda-forge 2026-08-19 21:32 UTC. hdmf 5.x and 6.1.0 were already published before Aug 18, so the exact flip is unpinned-solver drift rather than 6.2.0 alone. The hdmf version actually installed in the failing run is not recoverable from the logs: the backend log dump covers 01:00 to 01:01 only, and the run executed at 00:33.

Scope: only `caiman.yaml` is exposed today.

- `caiman.yaml`: `pynwb<3`, hdmf unpinned, python 3.11. Affected.
- `optinist.yaml`: `pynwb<3`, hdmf unpinned, but python 3.9, and hdmf >= 5 requires python >= 3.10, so it cannot pick up the breaking change until that env's python is bumped.
- `microscope.yaml`: pins `hdmf==3.*`. Safe.
- App image: `pyproject.toml` pins `hdmf = "3.*"` (3.14.6 installed). Safe.

### Measures

Add the missing upper bound to `studio/app/optinist/wrappers/caiman/conda/caiman.yaml`:

```diff
   - pynwb<3
+  - hdmf<4
```

The same dry-run solve with that pin added resolves cleanly, and to the same hdmf the app image already uses:

```
+ hdmf   3.14.6  pyhd8ed1ab_0  conda-forge
+ pynwb   2.8.3  pyhd8ed1ab_1  conda-forge
```

`hdmf<4` rather than `hdmf<5` because the app image reads these NWB files with hdmf 3.14.6, and `microscope.yaml` already pairs `pynwb==2.6.0` with `hdmf==3.*`. Keeping the writer on the same major as the reader matches the existing pattern.

Considered and rejected for now: migrating to pynwb 3 or 4 instead of pinning. pynwb 3.0.0 declares `hdmf >=3.14.5,<5` and pynwb 4.1.0 declares `hdmf >=6.1.0,<7`, so upstream added the ceiling that pynwb 2.x lacks, and a migration is version-solvable in the caiman env (pynwb 3 needs numpy >= 1.24; caiman 1.11.4 wants numpy >= 1.26.4,<2). But the app image and every env that writes NWB share these files, so the bump cannot be scoped to one env: it would move `pyproject.toml`, `caiman.yaml`, `optinist.yaml`, and `microscope.yaml` together, plus an audit of `nwb_creater.py` against pynwb 3's removed APIs and a revalidation of every tutorial's NWB output. It also does not address the defect that broke CI, which is an env with no version ceiling, so the ceiling is needed either way.

### Others

Consider adding the same `hdmf` ceiling to `optinist.yaml` pre-emptively, so a future python bump in that env does not reintroduce this.

Follow-up worth considering separately: the e2e job's "Dump backend logs on failure" step captured only the final minute of container output, so the node traceback had to be recovered from the Playwright trace instead of the backend log.

--- a/docs/specifications/add_algorithm.md
+++ b/docs/specifications/add_algorithm.md
@@ -141,6 +141,7 @@ dependencies:
 - Note:
   - Currently, Optinist packages supports NumPy versions below 2. Please ensure that your Conda environment is configured with an appropriate version of NumPy.
   - If you plan to use NumPy 2.x or higher, ensure all other packages in your environment are compatible with it.
+  - These envs are solved fresh at run time rather than baked into the image, so your node can get a different package set than the one you developed against, including transitive packages the yaml never lists. Pin your direct dependencies as tightly as the function allows.
 
 ### Check your custom node inputs and outputs
 

--- a/studio/app/optinist/wrappers/caiman/conda/caiman.yaml
+++ b/studio/app/optinist/wrappers/caiman/conda/caiman.yaml
@@ -10,5 +10,6 @@ dependencies:
   - pandas>=1.5.0
   - scikit-image>=0.19.0
   - pynwb<3
+  - hdmf<4
   - tensorflow>=2.12,<2.16
   - caiman>=1.11.3,<=1.11.4

--- a/studio/app/optinist/wrappers/optinist/conda/optinist.yaml
+++ b/studio/app/optinist/wrappers/optinist/conda/optinist.yaml
@@ -8,6 +8,7 @@ dependencies:
       - scikit-learn==1.1.*
       - statsmodels==0.13.*
       - pynwb<3
+      - hdmf<4
       - numexpr
       - numba
       - dpca

